### PR TITLE
Sign and place order

### PR DESCRIPTION
**Pull request by fleupold**
After having fetched the "random" token pair, we can now place an order for this pair, getting the fee and estimated amount information, signing the order and then placing it.

Note that orders are not yet tradable since we have not yet given approval to the Settlement contract. This will come in the next PR.

### CODE REVIEW

File: src/make_trade.ts [hunk 1]
```diff
@@ -15,28 +24,70 @@ async function makeTrade(
...
+    sellToken,
+    buyToken,
+    sellAmountAfterFee,
+    buyAmount,
+    fee
+  );
+  const signature = await signOrder(
+    domain(
+      network.config.chainId,
+      // TODO: How do I get this address from the repo?
```
Comment (fleupold) [inline src/make_trade.ts:80]: @nlordell could you help me with this TODO and the one below?

File: src/make_trade.ts [hunk 2]
```diff
@@ -15,28 +24,70 @@ async function makeTrade(
...
+      "0x4E608b7Da83f8E9213F554BDAA77C72e125529d0"
```
Comment (nlordell) [inline src/make_trade.ts:81]: For this, you could:
```ts
import { GPv2Settlement } from "@gnosis.pm/gp-v2-contracts/networks.json";

// ...

GPv2Settlement[network.config.chainId].address;
```

_disclaimer_: you may need to declare some types. I created an issue in the contracts repo to make this more convenient: https://github.com/gnosis/gp-v2-contracts/issues/545

File: src/make_trade.ts [hunk 3]
```diff
@@ -15,28 +24,70 @@ async function makeTrade(
...
+    ),
+    order,
+    trader,
+    // TODO: TypedData does not seem to work with the ethers provider
+    SigningScheme.MESSAGE
```
Comment (nlordell) [inline src/make_trade.ts:86]: In general, it does. I imagine the `trader` signer being used doesn't provide a `_TypedDataSigner` implementation.

File: src/make_trade.ts [hunk 4]
```diff
@@ -86,4 +137,36 @@ async function toERC20(
...
-export { makeTrade };
+// Using the most significant 4 bytes of a unique phrase's hash. TODO: use full hash after SC upgrade.
+const APP_DATA = parseInt(
+  ethers.utils.hexDataSlice(keccak("GPv2 Trading Bot"), 0, 4)
+);
+
+function keccak(message: string) {
+  const utf8Encoder = new TextEncoder();
+  return ethers.utils.keccak256(utf8Encoder.encode(message));
+}
```
Comment (nlordell) [inline src/make_trade.ts:148]:
```suggestion
const keccak = ethers.utils.id;
```

**Review by nlordell — COMMENTED** [review]
Just some small nits. Overall looks good.

File: src/make_trade.ts [hunk 5]
```diff
@@ -1,12 +1,21 @@
...
+  SigningScheme,
+  domain,
+} from "@gnosis.pm/gp-v2-contracts";
 import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
 import { HardhatEthersHelpers } from "@nomiclabs/hardhat-ethers/types";
 import ERC20 from "@openzeppelin/contracts/build/contracts/ERC20.json";
 import { TokenInfo, TokenList } from "@uniswap/token-lists";
-import { BigNumber, Contract } from "ethers";
+import { BigNumber, Contract, ethers } from "ethers";
 import { HardhatRuntimeEnvironment } from "hardhat/types";
```
Comment (nlordell) [inline src/make_trade.ts:13]:
```suggestion
import { BigNumber, Contract } from "ethers";
import { ethers } from "hardhat";
import { HardhatRuntimeEnvironment } from "hardhat/types";
```

I suggest using the `hardhat` wrapped `ethers` as it configures the provide and signer for you.

Comment (nlordell) [reply src/make_trade.ts:13]: **edit**: Looks like you can't do this for tasks.

File: src/make_trade.ts (hunk 3 above)
Comment (nlordell) [reply src/make_trade.ts:86]: Hmm, what exactly is not working? I tried this locally and I was able to sign.

File: src/api.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+  constructor(network: string) {
+    this.network = network;
+  }
+
+  async call<T>(route: string, init?: RequestInit): Promise<T> {
+    const url = `https://protocol-${this.network}.dev.gnosisdev.com/api/v1/${route}`;
+    const response = await fetch(url, init);
+    const body = await response.json();
+    if (Math.floor(response.status / 100) !== 2) {
```
Comment (alfetopito) [inline src/api.ts:16]: How many successful statuses do you expect to have? 203 and 204 are also ok?

Not a big deal, I just find this weird

File: src/make_trade.ts (hunk 2 above)
Comment (alfetopito) [reply src/make_trade.ts:81]: Example of getting that on GP-UI repo https://github.com/gnosis/gp-ui/blob/develop/src/utils/contract.ts

File: src/make_trade.ts (hunk 3 above)
Comment (alfetopito) [reply src/make_trade.ts:86]: Is this issue the same as in this TODO? https://github.com/gnosis/gp-swap-ui/blob/639d025a5ae1be189209e3639c248b4be2e46f79/src/custom/utils/signatures.ts#L25-L43

That typescript can't export enums?

File: src/api.ts (hunk 6 above)
Comment (nlordell) [reply src/api.ts:16]: I missed this, but maybe you can use [`response.ok`](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok).

File: src/make_trade.ts (hunk 3 above)
Comment (nlordell) [reply src/make_trade.ts]: I think the issue is the `hardhat` `ethers.Signer` implementation doesn't support `eth_signTypedData_v4` calls.

**Review by alfetopito — APPROVED** [review]

**Review by nlordell — APPROVED** [review]

**Pull request merged**